### PR TITLE
Reorganize plugin API around commands

### PR DESCRIPTION
**Pull request by devyn**
[Context on Discord](https://discord.com/channels/601130461678272522/855947301380947968/1216517833312309419)

# Description
This is a significant breaking change to the plugin API, but one I think is worthwhile. @ayax79 mentioned on Discord that while trying to start on a dataframes plugin, he was a little disappointed that more wasn't provided in terms of code organization for commands, particularly since there are *a lot* of `dfr` commands.

This change treats plugins more like miniatures of the engine, with dispatch of the command name being handled inherently, each command being its own type, and each having their own signature within the trait impl for the command type rather than having to find a way to centralize it all into one `Vec`.

For the example plugins that have multiple commands, I definitely like how this looks a lot better. This encourages doing code organization the right way and feels very good.

For the plugins that have only one command, it's just a little bit more boilerplate - but still worth it, in my opinion.

The `Box<dyn PluginCommand<Plugin = Self>>` type in `commands()` is a little bit hairy, particularly for Rust beginners, but ultimately not so bad, and it gives the desired flexibility for shared state for a whole plugin + the individual commands.

# User-Facing Changes
Pretty big breaking change to plugin API, but probably one that's worth making.

```rust
use nu_plugin::*;
use nu_protocol::{PluginSignature, PipelineData, Type, Value};

struct LowercasePlugin;
struct Lowercase;

// Plugins can now have multiple commands
impl PluginCommand for Lowercase {
    type Plugin = LowercasePlugin;

    // The signature lives with the command
    fn signature(&self) -> PluginSignature {
        PluginSignature::build("lowercase")
            .usage("Convert each string in a stream to lowercase")
            .input_output_type(Type::List(Type::String.into()), Type::List(Type::String.into()))
    }

    // We also provide SimplePluginCommand which operates on Value like before
    fn run(
        &self,
        plugin: &LowercasePlugin,
        engine: &EngineInterface,
        call: &EvaluatedCall,
        input: PipelineData,
    ) -> Result<PipelineData, LabeledError> {
        let span = call.head;
        Ok(input.map(move |value| {
            value.as_str()
                .map(|string| Value::string(string.to_lowercase(), span))
                // Errors in a stream should be returned as values.
                .unwrap_or_else(|err| Value::error(err, span))
        }, None)?)
    }
}

// Plugin now just has a list of commands, and the custom value op stuff still goes here
impl Plugin for LowercasePlugin {
    fn commands(&self) -> Vec<Box<dyn PluginCommand<Plugin=Self>>> {
        vec![Box::new(Lowercase)]
    }
}

fn main() {
    serve_plugin(&LowercasePlugin{}, MsgPackSerializer)
}
```

Time this however you like - we're already breaking stuff for 0.92, so it might be good to do it now, but if it feels like a lot all at once, it could wait.

# Tests + Formatting
- :green_circle: `toolkit fmt`
- :green_circle: `toolkit clippy`
- :green_circle: `toolkit test`
- :green_circle: `toolkit test stdlib`

# After Submitting
- [ ] Update examples in the book
- [x] Fix #12088 to match - this change would actually simplify it a lot, because the methods are currently just duplicated between `Plugin` and `StreamingPlugin`, but they only need to be on `Plugin` with this change

### CODE REVIEW

**Comment by devyn** [comment]
Updated for compatibility with #12088. As expected, it simplifies the interface a lot there

File: crates/nu-plugin/src/plugin/mod.rs [hunk 1]
```diff
@@ -581,83 +380,6 @@ pub trait StreamingPlugin: Sync {
...
-        &self,
-        engine: &EngineInterface,
-        custom_value: Box<dyn CustomValue>,
-    ) -> Result<(), LabeledError> {
-        <Self as Plugin>::custom_value_dropped(self, engine, custom_value)
-    }
-}
-
 /// Function used to implement the communication protocol between
 /// nushell and an external plugin. Both [Plugin] and [StreamingPlugin] are supported.
```
Comment (WindSoilder) [inline crates/nu-plugin/src/plugin/mod.rs:384]: Can you clean up some comments about `StreamingPlugin`?  I don't think it's required for now.

I can see there are three places use `StreamingPlugin`:
1. `crates/nu-plugin/src/plugin/mod.rs: line 384
2. `crates/nu-protocol/src/value/custom_value.rs: line 82
3. `crates/nu_plugin_stream_example/README.md: line 3

图片

File: crates/nu-plugin/src/plugin/mod.rs [hunk 2]
```diff
@@ -708,6 +428,13 @@ pub fn serve_plugin(plugin: &impl StreamingPlugin, encoder: impl PluginEncoder +
         std::process::exit(1)
     }
 
+    // Build commands map, to make running a command easier
+    let commands: HashMap<String, _> = plugin
+        .commands()
+        .into_iter()
+        .map(|command| (command.signature().sig.name.clone(), command))
```
Comment (WindSoilder) [inline crates/nu-plugin/src/plugin/mod.rs:435]: Not really sure: Should we output warning messages if there are two commands have the same name?

**Review by WindSoilder — COMMENTED** [review]
Thank you!  I really like the change!  It enables user writing plugin commands in a more modular way.

File: crates/nu-plugin/src/plugin/mod.rs (hunk 1 above)
Comment (devyn) [reply crates/nu-plugin/src/plugin/mod.rs:384]: Oh, good catch - I'll address these.

File: crates/nu-plugin/src/plugin/mod.rs (hunk 2 above)
Comment (devyn) [reply crates/nu-plugin/src/plugin/mod.rs:435]: I thought about that too. I think it's a good idea.

**Comment by devyn** [comment]
@WindSoilder latest changes should address your feedback. Thanks!

**Comment by fdncred** [comment]
Thanks @devyn!

**Pull request merged**